controlplane/kubeadm/internal/workload_cluster.go

```diff
@@ -518,35 +512,27 @@ func (w *Workload) ForwardEtcdLeadership(ctx context.Context, machine *clusterv1
 		return nil
 	}
 
-	// TODO we'd probably prefer to pass in all the known nodes and let grpc handle retrying connections across them
-	clientMachineName := machine.Status.NodeRef.Name
-	if leaderCandidate != nil && leaderCandidate.Status.NodeRef != nil {
-		// connect to the new leader candidate, in case machine's etcd membership has already been removed
-		clientMachineName = leaderCandidate.Status.NodeRef.Name
+	nodes, err := w.getControlPlaneNodes(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to list control plane nodes")
 	}
 
-	etcdClient, err := w.etcdClientGenerator.forNode(ctx, clientMachineName)
+	etcdClient, err := w.etcdClientGenerator.forLeader(ctx, nodes)
 	if err != nil {
 		return errors.Wrap(err, "failed to create etcd Client")
 	}
 
-	// List etcd members. This checks that the member is healthy, because the request goes through consensus.
 	members, err := etcdClient.Members(ctx)
 	if err != nil {
 		return errors.Wrap(err, "failed to list etcd members using etcd client")
 	}
 
 	currentMember := etcdutil.MemberForName(members, machine.Status.NodeRef.Name)
 	if currentMember == nil || currentMember.ID != etcdClient.LeaderID {
+		// nothing to do, this is not the etcd leader
 		return nil
 	}
 
-	// Move the etcd client to the current leader, which in this case is the machine we're about to delete.
-	etcdClient, err = w.etcdClientGenerator.forNode(ctx, machine.Status.NodeRef.Name)
-	if err != nil {
-		return errors.Wrap(err, "failed to create etcd Client")
-	}
-
 	// If we don't have a leader candidate, move the leader to the next available machine.
 	if leaderCandidate == nil || leaderCandidate.Status.NodeRef == nil {
```

Comment:
Let's also remove this check, and make `leaderCandidate` a required parameter, otherwise return an error